cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -657,34 +986,85 @@ class AsofJoinNode : public ExecNode {
     return std::make_shared<arrow::Schema>(fields);
   }
 
+  static inline Result<col_index_t> FindColIndex(const Schema& schema,
+                                                 const FieldRef& field_ref,
+                                                 const util::string_view& key_kind) {
+    auto match_res = field_ref.FindOne(schema);
+    if (!match_res.ok()) {
+      return Status::Invalid("Bad join key on table : ", match_res.status().message());
+    }
+    auto match = match_res.ValueOrDie();
+    if (match.indices().size() != 1) {
+      return Status::Invalid("AsOfJoinNode does not support a nested ",
+                             to_string(key_kind), "-key ", field_ref.ToString());
+    }
+    return match.indices()[0];
+  }
+
   static arrow::Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
                                        const ExecNodeOptions& options) {
     DCHECK_GE(inputs.size(), 2) << "Must have at least two inputs";
 
     const auto& join_options = checked_cast<const AsofJoinNodeOptions&>(options);
-    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Schema> output_schema,
-                          MakeOutputSchema(inputs, join_options));
+    if (join_options.by_key.size() == 0) {
+      return Status::Invalid("AsOfJoin by_key must not be empty");
+    }
```

Comment:
Is that true?  For some reason I thought the by key was optional and the on key was the only required key.

Reply:
Good question. It wasn't optional prior to this change; only a single-by-key was accepted. Now that a multi-by-key is accepted, we could make it optional. @icexelloss, your thoughts?

Reply:
I've looked into this and it is not a trivial change, not only because new expected results (the empty key means all 3 joined tables will contribute a "key" column) need to be validated but also because I ran into unclear runtime problems on my first couple of tries. So, I'll need some time, or we could decide to split this into a separate issue.

Reply:
I managed to get this done in the latest commit here. I noticed that `Pandas.asof_merge`, for example, appears to accept an empty key, so `AsofJoin` should too.

Reply:
I think splitting into a separate issue is fine.

Reply:
As noted, I ended up fixing this in the latest commit in this PR, so no need to split to a separate issue.